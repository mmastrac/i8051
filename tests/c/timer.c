/* timer.c â€” 8051 timer test program */

#include <8051.h>  /* gives you SFR definitions like P1 */

volatile int interrupt_flag = 0;

void timer0_isr(void) __interrupt (1)
{
    /* Toggle port 1 and stop timer 0 */
    TR0 = 0;
    P1 = ~P1;
    interrupt_flag = 1;
}

void main(void)
{
    unsigned char i, sum = 0;
    __xdata unsigned char *scratch = (__xdata unsigned char *)0x8000;

    TMOD = 5; // 0b0000_0101: timer 0: mode 1 counter, 16-bit timer, auto-reload
    TH0 = 0xff;
    TL0 = 0xf4;
    ET0 = 1; // enable timer 0 interrupt
    EA = 1; // enable global interrupts
    TR0 = 1; // start timer 0

    while (!interrupt_flag) {
    }

    interrupt_flag = 0;

    TMOD = 1; // 0b0000_0001: timer 0: mode 1 timer, 16-bit timer, auto-reload
    TH0 = 0xff;
    TL0 = 0xf0;
    TR0 = 1; // start timer 0

    while (!interrupt_flag) {
    }

    /* Stop here (endless loop) */
    while (1) {
        /* idle */
    }
}
