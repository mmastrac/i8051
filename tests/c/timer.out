PORT read A8
PORT write A8 = 02
PORT read A8
PORT write A8 = 82
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read B0
PORT read latch 90
PORT write 90 = FF
PORT read latch 90
PORT write 90 = 00
CPU halted at 0x00F2 after 615 instructions
  A=01  B=00  DPTR=0000  C=false OV=false AC=false F0=false
  R0=00 R1=00 R2=00 R3=00 R4=00 R5=00 R6=00 R7=00 
MEM: 00 00 00 00
